plutus-core/src/Language/PlutusCore/Flat.hs

```diff
@@ -0,0 +1,356 @@
+    encode (Some (ValueOf uni x)) = encode (Some $ TypeIn uni) <> bring (Proxy @Flat) uni (encode x)
+
+    decode = go =<< decode where
+        go (Some (TypeIn uni)) = Some . ValueOf uni <$> bring (Proxy @Flat) uni decode
+
+    -- We need to get the flat instance in scope.
+    size (Some (ValueOf uni x)) acc = size (Some $ TypeIn uni) acc
+                                        + bring (Proxy @Flat) uni (size x 0)
+
+-- | Using 5 bits to encode builtin tags.
+builtinTagWidth :: NumBits
+builtinTagWidth = 5
+
+encodeBuiltin :: Word8 -> Encoding
+encodeBuiltin = eBits builtinTagWidth
+
+decodeBuiltin :: Get Word8
+decodeBuiltin = dBEBits8 builtinTagWidth
+
+instance Flat StaticBuiltinName where
+    encode = encodeBuiltin . \case
+              AddInteger           -> 0
+              SubtractInteger      -> 1
+              MultiplyInteger      -> 2
+              DivideInteger        -> 3
+              RemainderInteger     -> 4
+              LessThanInteger      -> 5
+              LessThanEqInteger    -> 6
+              GreaterThanInteger   -> 7
+              GreaterThanEqInteger -> 8
+              EqInteger            -> 9
+              Concatenate          -> 10
+              TakeByteString       -> 11
+              DropByteString       -> 12
+              SHA2                 -> 13
+              SHA3                 -> 14
+              VerifySignature      -> 15
+              EqByteString         -> 16
+              QuotientInteger      -> 17
+              ModInteger           -> 18
+              LtByteString         -> 19
+              GtByteString         -> 20
+              IfThenElse           -> 21
+
+    decode = go =<< decodeBuiltin
+        where go 0  = pure AddInteger
+              go 1  = pure SubtractInteger
+              go 2  = pure MultiplyInteger
+              go 3  = pure DivideInteger
+              go 4  = pure RemainderInteger
+              go 5  = pure LessThanInteger
+              go 6  = pure LessThanEqInteger
+              go 7  = pure GreaterThanInteger
+              go 8  = pure GreaterThanEqInteger
+              go 9  = pure EqInteger
+              go 10 = pure Concatenate
+              go 11 = pure TakeByteString
+              go 12 = pure DropByteString
+              go 13 = pure SHA2
+              go 14 = pure SHA3
+              go 15 = pure VerifySignature
+              go 16 = pure EqByteString
+              go 17 = pure QuotientInteger
+              go 18 = pure ModInteger
+              go 19 = pure LtByteString
+              go 20 = pure GtByteString
+              go 21 = pure IfThenElse
+              go _  = fail "Failed to decode BuiltinName"
+
+    size _ n = n + 5
+
+instance Flat Unique where
+    encode (Unique i) = eInt i
+    decode = Unique <$> dInt
+    -- There is no Generic instance for Unique,
+    -- so a `size` function cannot be generated.
+    size (Unique i) = sInt i
+
+instance Flat Name where
+    -- TODO: should we encode the name or not?
```

Comment:
Yes. We have the debruijn version for when we don't want to.

Reply:
It was already encoding them, I removed the comment.